services/galley/src/Galley/API/Federation.hs

```diff
@@ -529,26 +533,58 @@ instance
 
 mlsSendWelcome ::
   Members
-    '[ GundeckAccess,
+    '[ BrigAccess,
+       GundeckAccess,
        Input (Local ()),
        Input UTCTime
      ]
     r =>
   Domain ->
   F.MLSWelcomeRequest ->
-  Sem r ()
-mlsSendWelcome _origDomain (F.MLSWelcomeRequest b64RawWelcome rcpts) = do
+  Sem r F.MLSWelcomeResponse
+mlsSendWelcome _origDomain (F.unMLSWelcomeRequest -> welBase64) = do
   loc <- input @(Local ())
   now <- input @UTCTime
-  let rawWelcome = fromBase64ByteString b64RawWelcome
-  void $
-    runMessagePush loc Nothing $
-      foldMap (uncurry $ mkPush rawWelcome loc now) (F.unMLSWelRecipient <$> rcpts)
+  let rawWelcome = fromBase64ByteString welBase64
+  runErrors $ do
+    welcome <-
+      fromEither
+        . mapLeft (const . mlsProtocolError $ decodingErrorMsg)
+        . decodeMLS'
+        $ rawWelcome
+    -- Extract only recipients local to this backend
+    rcpts <-
+      fmap fold $
+        foldQualified loc ((: []) . tUnqualified) (const mempty)
+          <$$> welcomeRecipients welcome
+    void $
+      runMessagePush loc Nothing $
+        foldMap (uncurry $ mkPush rawWelcome loc now) rcpts
```

Comment:
Why is this copy-pasted from `Galley.API.MLS.Welcome` instead of simply using `sendLocalWelcomes`?

Reply:
We've refactored this yesterday while pairing to reuse `sendLocalWelcomes`.